clients/venice-thin-client/src/main/java/com/linkedin/venice/client/store/streaming/AbstractRecordStreamDecoder.java

```diff
@@ -0,0 +1,239 @@
+
+
+/**
+ * Streaming callback for batch-get/compute.
+ *
+ * Since data chunk returned by {@link D2TransportClient} doesn't respect chunk-size associated with each
+ * chunk sent by Venice Router, here is leveraging {@link ReadEnvelopeChunkedDeserializer} to deserialize
+ * data chunks even with the data chunk contains partial record in a non-blocking way.
+ *
+ * The envelope deserialization will happen in TransportClient thread pool (R2 thread pool for example if using
+ * {@link D2TransportClient}, and both the record deserialization and application's callback will be executed in
+ * Venice thread pool: {@link #deserializationExecutor},
+ *
+ * @param <ENVELOPE>
+ * @param <K>
+ * @param <V>
+ */
+public abstract class AbstractRecordStreamDecoder<ENVELOPE, K, V> implements RecordStreamDecoder {
+  private final List<K> keyList;
+  private final TrackingStreamingCallback<K, V> callback;
+  private final List<CompletableFuture<Void>> deserializationFutures = new ArrayList<>();
+  private final BitSet receivedKeySet; // Track which key is present in the response
+  private final AtomicInteger successfulKeyCnt = new AtomicInteger(0);
+  private final long preSubmitTimeInNS;
+  private final LongAdder deserializationTimeInNS = new LongAdder();
+  private final Executor deserializationExecutor;
+
+  // non-final state
+  private int duplicateEntryCount = 0;
+  private Optional<StreamingFooterRecordV1> streamingFooterRecord = Optional.empty();
+  private boolean isStreamingResponse = false;
+  private int responseSchemaId = SchemaData.INVALID_VALUE_SCHEMA_ID;
+  private CompressionStrategy compressionStrategy = CompressionStrategy.NO_OP;
+  private ReadEnvelopeChunkedDeserializer<ENVELOPE> envelopeDeserializer = null;
+
+  public AbstractRecordStreamDecoder(
+      List<K> keyList,
+      TrackingStreamingCallback<K, V> callback,
+      Executor deserializationExecutor) {
+    this.keyList = keyList;
+    this.callback = callback;
+    this.deserializationExecutor = deserializationExecutor;
+    this.preSubmitTimeInNS = System.nanoTime();
+    this.receivedKeySet = new BitSet(keyList.size());
+  }
+
+  @Override
+  public void onHeaderReceived(Map<String, String> headers) {
+    callback.getStats()
+        .ifPresent(
+            stats -> stats
+                .recordRequestSubmissionToResponseHandlingTime(LatencyUtils.getLatencyInMS(preSubmitTimeInNS)));
+    isStreamingResponse = headers.containsKey(HttpConstants.VENICE_STREAMING_RESPONSE);
+    String schemaIdHeader = headers.get(HttpConstants.VENICE_SCHEMA_ID);
+    if (schemaIdHeader != null) {
+      responseSchemaId = Integer.parseInt(schemaIdHeader);
+    }
+    envelopeDeserializer = getEnvelopeDeserializer(responseSchemaId);
+    String compressionHeader = headers.get(HttpConstants.VENICE_COMPRESSION_STRATEGY);
+    if (compressionHeader != null) {
+      compressionStrategy = CompressionStrategy.valueOf(Integer.parseInt(compressionHeader));
```

Comment:
can we handle illegal argument exception from `valueOf` here ?

Reply:
The exception is caught in `TransportClient::streamPost`, so there is no gap here. In terms of handling, not much we can do if we get data than cannot be understood besides from failing the request.

Reply:
I believe failing the request in that case is good idea?

Reply:
That's exactly what happens in this case, any unhandled exception fail the request.

Reply:
One clarification, `AbstractRecordStreamDecoder` is not exactly a new code. I copied it almost as-is from `AbstractAvroStoreClient` as part of refactoring.